### PR TITLE
Fix strict checkpoint loading with idle rerun state

### LINKED ISSUE #7428: [BUG] Strict loading rejects checkpoints with idle rerun state
**Describe the bug**

A normal `torch_dist` checkpoint saved with rerun validation enabled can fail to load with `dist_ckpt_strictness=raise_all`, even when the parallel topology is unchanged and no rerun is pending.

The saved checkpoint contains an idle `rerun_state_machine_state` shard. `load_checkpoint()` omits that shard from the load request because `validate_state_dict()` rejects the idle state as a rerun to restore. Strict key validation then fails:

```text
Missing keys (for all ranks): {'rerun_state_machine_state'}.
```

**Steps/Code to reproduce bug**

1. On two GPUs, save a `torch_dist` checkpoint with the rerun machine in `RerunMode.VALIDATE_RESULTS` and `RerunState.NOT_RUNNING_YET`.
2. Load it with the same topology, `dist_ckpt_strictness='raise_all'`, and `finetune=False`. Optimizer and RNG save/load can be disabled to isolate the issue.
3. The load fails with the missing-key error above. This reproduces with either `DISABLED` or `VALIDATE_RESULTS` as the loading job's rerun mode.

The minimal save/load test is [`test_strict_load_idle_rerun_checkpoint`](https://github.com/NVIDIA/Megatron-LM/blob/64a7a1ac37c9b7973f551d7efd6c3427b78258a1/tests/unit_tests/test_checkpointing.py#L511) in #7222. In a two-GPU source comparison, both cases fail with `checkpointing.py` from `7613fd7a9508833d5895dd35d2847a1b8571fd0d` and pass with the proposed fix.

**Expected behavior**

Load the checkpoint without relaxing strict key validation or starting a rerun. Model weights, iteration and FLOP counters should be restored, and the loading job's rerun mode should be preserved.

**Additional context**

Proposed fix: #7222 includes the idle shard in eligible load requests while keeping it excluded from rerun execution.

Related to #4448 and #4378. Those addressed the save-side structure of fault checkpoints; this report covers strict loading of an ordinary checkpoint containing the idle sentinel.

@NVIDIA/mcore-oncall

### PR DESCRIPTION
**Pull request by bzantium**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Fix strict loading of checkpoints saved with idle rerun state.

Normal checkpoints saved with rerun validation enabled contain an idle sentinel so cached save plans stay consistent. The load path currently omits that shard because idle state is not an active rerun, causing `raise_all` strictness to report a missing key.

Include the saved shard in eligible same-topology load requests while keeping it excluded from rerun execution. Strict key validation and the existing release, finetune, and topology checks remain unchanged.

## Testing

Both two-GPU `torch_dist` save/load cases pass after rebasing onto `8d5555c93`, covering enabled and disabled rerun validation in the loading job. They check restored model weights, iteration and FLOP counters, and that the loading job's rerun mode is preserved without starting a rerun. Both cases fail with the missing-shard error when using `checkpointing.py` from that base revision.

The test file passes Black. Full-file formatting reports existing differences in `checkpointing.py` outside this change. NVIDIA CI is still pending maintainer vetting.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: Fixes #7428. Related to #4448 and #4378.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
- Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing. Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by bzantium** [comment]
Could a maintainer vet this PR so the NVIDIA CI can run? The two-GPU save/load regressions pass locally for both enabled and disabled rerun validation; both reproduce the missing-shard error on the base revision. The branch is still in draft pending CI.
